### PR TITLE
fix: list object by bucket name bug

### PR DESCRIPTION
**Pull request by BarryTong65**
### Description

Fix the bug where deleting all objects within a folder at the same block height does not result in the folder being deleted

### Rationale

image

When a user deletes the objects beneath the /coco/ folder, leaving it empty, it should be deleted according to the block syncer logic. However, it is unexpectedly retained, resulting in a deleted folder "/coco/" appearing in the user's bucket.

```
"/coco/"
"/coco/data/train-00000-of-00040-67e35002d152155c.parquet"
"/coco/data/train-00004-of-00040-1df1755d6f7c9b17.parquet"
"/coco/data/train-00003-of-00040-a77e00648e423963.parquet"
"/coco/data/train-00001-of-00040-2c2b33b9504aa843.parquet"
"/coco/data/train-00002-of-00040-58e30306870b1de7.parquet"
```

image
https://greenfieldscan.com/bucket/0x00000000000000000000000000000000000000000000000000000000000007d8
When querying the user's bucket information on the browser side, all objects related to "/coco/" have disappeared, indicating that the bug resides here.
image

Root Cause:
The issue arises because subsequent optimizations in the block syncer aggregate all SQL queries within the same block height to execute together. Within a single block height, if I delete all objects under "/coco/", the deletion logic is processed, but the SQL is not executed yet. When it comes to deleting the last object, "/coco/data/train-00002-of-00040-58e30306870b1de7.parquet", the database still shows 5 objects under the "/coco/" folder (in reality, only the last one remains, and according to serial logic, "/coco/" should be deleted). As a result, both "/coco/" and "/coco/data/" folders are not deleted as they should be.

### Example

N/A

### Changes

Notable changes: 
* By creating a global context in the block syncer module and using WithValue to implement operations similar to a map, query the delete object and create object events within the current block height to avoid the aforementioned bug.
* "mplement a script to delete erroneous data in the database caused by multiple deletions or creations of an object within the same block height.
image

### Potential Impacts
* QA Test @cd1129678576

### CODE REVIEW

File: modular/blocksyncer/blocksyncer_options.go [hunk 1]
```diff
@@ -600,3 +602,191 @@ func (b *BlockSyncerModular) syncBucketSize() error {
...
+			})
+		}
+		if len(nodes) == 0 {
+			continue
+		}
+		// 1. First, check if all sub-folders also contain only one folder and no objects. If true, delete them and proceed to check the parent folder logic.
+		// 2. For parent folders, check if they too contain only one folder and no objects. Recursively perform this check up to the root ('/') folder, and then directly delete it.
+		// 3. End this loop.
+		for _, node := range nodes {
+			// 这里使用full name的原因是看看这个文件夹下面是不为空
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply modular/blocksyncer/blocksyncer_options.go]: fixed

File: modular/blocksyncer/modules/prefixtree/prefix_tree.go [hunk 2]
```diff
@@ -202,7 +279,20 @@ func (m *Module) deleteObject(ctx context.Context, objectPath, bucketName string
 		return nil, nil
 	}
 	k, v := m.db.DeletePrefixTree(ctx, nodes)
+	m.SetCtx(GetObjectNameExist(objectID.String()), 0)
 	return map[string][]interface{}{
 		k: v,
 	}, nil
 }
+
+func GetFolderNameExist(folder string) string {
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply modular/blocksyncer/modules/prefixtree/prefix_tree.go]: fixed, replaced with GenerateFolderExistKey/GenerateObjectExistKey/GenerateFolderCountKey

File: modular/blocksyncer/modules/prefixtree/prefix_tree.go [hunk 3]
```diff
@@ -202,7 +279,20 @@ func (m *Module) deleteObject(ctx context.Context, objectPath, bucketName string
...
+	return fmt.Sprintf("slash_prefix_tree:folder:exist:%s", folder)
+}
+
+func GetObjectNameExist(object string) string {
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply modular/blocksyncer/modules/prefixtree/prefix_tree.go]: fixed, replaced with GenerateFolderExistKey/GenerateObjectExistKey/GenerateFolderCountKey

File: modular/blocksyncer/modules/prefixtree/prefix_tree.go [hunk 4]
```diff
@@ -202,7 +279,20 @@ func (m *Module) deleteObject(ctx context.Context, objectPath, bucketName string
...
+	return fmt.Sprintf("slash_prefix_tree:object:exist:%s", object)
+}
+
+func GetFolderNameCount(folder string) string {
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply modular/blocksyncer/modules/prefixtree/prefix_tree.go]: fixed, replaced with GenerateFolderExistKey/GenerateObjectExistKey/GenerateFolderCountKey

File: modular/blocksyncer/modules/prefixtree/prefix_tree.go [hunk 5]
```diff
@@ -99,46 +100,110 @@ func (m *Module) handleCreateObject(ctx context.Context, sealObject *storagetype
...
-			// If the directory does not exist, create it
-			newNode := &bsdb.SlashPrefixTreeNode{
-				PathName:   strings.Join(pathParts[:i-1], "/") + "/",
-				FullName:   path,
-				Name:       pathParts[i-1] + "/",
-				IsObject:   false,
-				IsFolder:   true,
-				BucketName: bucketName,
-				ObjectName: "",
+		// Update how many object exists in under this folder
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply modular/blocksyncer/modules/prefixtree/prefix_tree.go:152]: updated

**Review by ruojunm — APPROVED** [review]
LGTM

**Review by constwz — APPROVED** [review]

**Pull request merged**